// @mercurial-finance/dynamic-amm-sdk/dist/cjs/src/amm/tests/utils/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEVNET = exports.MAINNET = exports.mockWallet = exports.airDropSol = void 0;
const anchor_1 = require("@project-serum/anchor");
const bytes_1 = require("@project-serum/anchor/dist/cjs/utils/bytes");
const web3_js_1 = require("@solana/web3.js");
const LAMPORTS_PER_SOL = 1e9;
const airDropSol = (connection, publicKey, amount = 1) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const airdropSignature = yield connection.requestAirdrop(publicKey, amount * LAMPORTS_PER_SOL);
        const latestBlockHash = yield connection.getLatestBlockhash();
        yield connection.confirmTransaction({
            blockhash: latestBlockHash.blockhash,
            lastValidBlockHeight: latestBlockHash.lastValidBlockHeight,
            signature: airdropSignature,
        });
    }
    catch (error) {
        console.error(error);
        throw error;
    }
});
exports.airDropSol = airDropSol;
exports.mockWallet = new anchor_1.Wallet(process.env.WALLET_PRIVATE_KEY ? web3_js_1.Keypair.fromSecretKey(bytes_1.bs58.decode(process.env.WALLET_PRIVATE_KEY)) : new web3_js_1.Keypair());
exports.MAINNET = {
    connection: new web3_js_1.Connection(process.env.MAINNET_RPC_ENDPOINT),
    cluster: 'mainnet-beta',
};
exports.DEVNET = {
    connection: new web3_js_1.Connection('https://api.devnet.solana.com/', {
        commitment: 'confirmed',
    }),
    cluster: 'devnet',
};
